src/java.base/share/classes/java/security/PEMDecoder.java

```diff
@@ -0,0 +1,427 @@
+                        X509EncodedKeySpec spec = new X509EncodedKeySpec(
+                            p8key.getPubKeyEncoded(), algo);
+                        yield new KeyPair(getKeyFactory(algo).
+                            generatePublic(spec), (PrivateKey) d);
+
+                    } else if (d instanceof PKCS8Key p8 &&
+                        p8.getPubKeyEncoded() != null) {
+                        // If the KeyFactory decoded an algorithm-specific
+                        // encodings, look for the public key again.  This
+                        // happens with EC and SEC1-v2 encoding
+                        X509EncodedKeySpec spec = new X509EncodedKeySpec(
+                            p8.getPubKeyEncoded(), algo);
+                        yield new KeyPair(getKeyFactory(algo).
+                            generatePublic(spec), p8);
+                    } else {
+                        // No public key, return the private key.
+                        yield d;
+                    }
+                }
+                case Pem.ENCRYPTED_PRIVATE_KEY -> {
+                    if (password == null) {
+                        yield new EncryptedPrivateKeyInfo(decoder.decode(
+                            pem.pem()));
+                    }
+                    yield new EncryptedPrivateKeyInfo(decoder.decode(pem.pem())).
+                        getKey(password.getPassword());
+                }
+                case Pem.CERTIFICATE, Pem.X509_CERTIFICATE -> {
+                    CertificateFactory cf = getCertFactory("X509");
+                    yield (X509Certificate) cf.generateCertificate(
+                        new ByteArrayInputStream(decoder.decode(pem.pem())));
+                }
+                case Pem.X509_CRL -> {
+                    CertificateFactory cf = getCertFactory("X509");
+                    yield (X509CRL) cf.generateCRL(
+                        new ByteArrayInputStream(decoder.decode(pem.pem())));
+                }
+                case Pem.RSA_PRIVATE_KEY -> {
+                    KeyFactory kf = getKeyFactory("RSA");
+                    yield kf.generatePrivate(
+                        RSAPrivateCrtKeyImpl.getKeySpec(decoder.decode(
+                            pem.pem())));
+                }
+                default -> pem;
+            };
+        } catch (GeneralSecurityException | IOException e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    /**
+     * Decodes and returns a {@link DEREncodable} from the given {@code String}.
+     *
+     * <p>This method reads the {@code String} until PEM data is found or until
+     * the end is reached.  Non-PEM data before the PEM header is ignored by
+     * the decoder.
+     *
+     * @param str a String containing PEM data.
+     * @return a {@code DEREncodable} generated from the PEM data.
+     * @throws IllegalArgumentException on error in decoding or if the PEM is
+     * unsupported.
```

Comment:
If the PEM is unsupported, you return a `PEMRecord` now, so you can remove those words. Same comment on lines 248-249.